fv3core/utils/grid.py

```diff
@@ -366,14 +367,203 @@ def horizontal_starts_from_shape(self, shape):
+    Provides indices for cell-centered variables with halos.
+
+    These indices can be used with horizontal interface variables by adding 1
+    to the domain shape along any interface axis.
+    """
+
+    def __init__(
+        self,
+        origin: Index3D,
+        domain: Index3D,
+        n_halo: int,
+        south_edge: bool,
+        north_edge: bool,
+        west_edge: bool,
+        east_edge: bool,
+    ):
+        """
+        Initialize a grid indexing object.
+
+        Args:
+            origin: index of the compute domain origin for cell-centered variables
+            domain: size of the compute domain for cell-centered variables
+            n_halo: number of halo points
+            south_edge: whether the current rank is on the south edge of a tile
+            north_edge: whether the current rank is on the north edge of a tile
+            west_edge: whether the current rank is on the west edge of a tile
+            east_edge: whether the current rank is on the east edge of a tile
+        """
+        self.origin = origin
+        self.domain = domain
+        self.n_halo = n_halo
+        # TODO: split edge booleans into their own class, they aren't needed
+        # for any of the indexing operations, only needed for axis_offsets
+        # and they make the API pretty big
+        # maybe give this responsibility to CubedSphereCommunicator
+        # as it already contains this information
+        self.south_edge = south_edge
+        self.north_edge = north_edge
+        self.west_edge = west_edge
+        self.east_edge = east_edge
+
+    @classmethod
+    def from_sizer_and_communicator(
+        cls, sizer: fv3util.GridSizer, cube: fv3util.CubedSphereCommunicator
+    ) -> "GridIndexing":
+        # TODO: if this class is refactored to split off the *_edge booleans,
+        # this init routine can be refactored to require only a GridSizer
+        origin = sizer.get_origin([fv3util.X_DIM, fv3util.Y_DIM, fv3util.Z_DIM])
+        domain = sizer.get_extent([fv3util.X_DIM, fv3util.Y_DIM, fv3util.Z_DIM])
+        south_edge = cube.tile.on_tile_bottom(cube.rank)
+        north_edge = cube.tile.on_tile_top(cube.rank)
+        west_edge = cube.tile.on_tile_left(cube.rank)
+        east_edge = cube.tile.on_tile_right(cube.rank)
+        return cls(
+            origin=origin,
+            domain=domain,
+            n_halo=sizer.n_halo,
+            south_edge=south_edge,
+            north_edge=north_edge,
+            west_edge=west_edge,
+            east_edge=east_edge,
+        )
+
+    @classmethod
+    def from_legacy_grid(cls, grid: Grid) -> "GridIndexing":
+        return cls(
+            origin=grid.compute_origin(),
+            domain=grid.domain_shape_compute(),
+            n_halo=grid.halo,
+            south_edge=grid.south_edge,
+            north_edge=grid.north_edge,
+            west_edge=grid.west_edge,
+            east_edge=grid.east_edge,
+        )
+
+    @property
+    def isc(self):
```

Comment:
it is strange to me that the origin and domain initializing this object can be whatever, and the indices then are relative to that provided origin and domain. The compute domain should be the same indices always in this model. It might be more clear to pass in the 'c' value, e.g. 12, 48, whatever, and then set all of these attributes base on that. I am open to discussion on this. Is there a potential future use case where we'd want to pass in a different origin and domain to this object?

Reply:
We currently have a hard-codable compute origin and compute domain, but it's very possible we'd want to change that in the future, e.g. to use 4 halos for certain variables. The Sizer class currently does the conversion from cXX to an origin and domain, so we already have that initialization responsibility covered.

Reply:
I don't know this use case of wanting to use 4 halos for certain variables -- what modules/experiments are you thinking of that would use this? The GridIndexing class then could use the Sizer class. This new class implies more flexibility than is possible to work(the old class does too, but while we are rewriting it, good to consider).

Reply:
> This new class implies more flexibility than is possible to work

Is that true? If we use the GridIndexing everywhere to handle horizontal dimensions, it should "just work" to provide arrays with 4 halo points to fv3core (alongside the appropriate GridIndexing instance). It's also hard to be certain we don't need to support 4 halo points, because fv3core isn't the top level of the model. If we hard-code 3 halo points, an ML team can't make physics schemes or integrate analysis code that needs more halo points.

Probably more importantly, it's a fairly complicated responsibility to go from a configuration to the shape of your arrays on a given rank. You need not just the cXX value, but also the layout, the rank, and how data is distributed on a tile across ranks. All of this is encapsulated right now [in Sizer](https://github.com/VulcanClimateModeling/fv3gfs-util/blob/master/fv3gfs/util/initialization/sizer.py#L32) (which in turn uses TilePartitioner for most of the work).

Reply:
What I mean is that the compute origin and domain should be fixed for all of the current model, so there is flexibility that is not currently needed. If there is indeed a known use case, sure, we can provide that.  Do some of the physics schemes or ML code actually need more halo points? I haven't heard of anyone wanting this use case in this model, but of course doesn't mean it won't be a possible request this would be ready for. The local index values set and used here should not depend on the rank and layout (though the 'edge' variables currently in the class will for now). The job of this class is relatively simple, I just hope it doesn't end up getting used in weird ways (like making a different grid indexing object of staggered variables), such that the meaning of the compute start and end indices changes dramatically. This is probably fine for now.

Reply:
> The local index values set and used here should not depend on the rank and layout (though the 'edge' variables currently in the class will for now).

We're thinking pretty seriously about reducing the number of points given to ranks at tile edges, to even out the work since they have extra edge handling logic to perform. This would make these local index values depend on rank.

The local index values already depend on layout. End indices are half as big for a (2, 2) layout as for a (1, 1) layout.

> The job of this class is relatively simple, I just hope it doesn't end up getting used in weird ways (like making a different grid indexing object of staggered variables)

Yes, let's try to avoid that. The new `origin_domain([X_DIM, Y_INTERFACE_DIM])` should hopefully placate anyone trying to handle interface variables.